### ISSUE TITLE
Dynamic SPN example

### ISSUE THREAD

**Issue by anicolson** [pronobis/libspn-keras#11]
Hi Jos,

I notice that you have started implementing dynamic SPNs:
https://github.com/pronobis/libspn-eras/blob/master/libspn_keras/layers/temporal_dense_product.py

Thats great!

I was wondering if you could give a simple tutorial about how to implement a dynamic SPN.

Thanks.

**Comment by jostosh** [comment]
Hi Aaron, I'm definitely planning to do this. Any suggestions for a dataset?

**Comment by anicolson** [comment]
How about something simple like sine wave mapping:

Observation: sine wave.
Target: sine wave 180 degrees out of phase.

We can use the positional encoding to the Transformer network as the feature (where each feature is a sine wave with a different frequency) and take its negative as the target:

```
import numpy as np
import tensorflow as tf

## POSITIONAL ENCODING
#
## NOTE: this implementation is from http://jalammar.github.io/illustrated-transformer/
## 		 and has been modified for TensorFlow 2.x.
#
def get_timing_signal_1d(length,
	channels,
	min_timescale=1.0,
	max_timescale=1.0e4,
	start_index=0):

	"""Gets a bunch of sinusoids of different frequencies.
	Each channel of the input Tensor is incremented by a sinusoid of a different
	frequency and phase.
	This allows attention to learn to use absolute and relative positions.
	Timing signals should be added to some precursors of both the query and the
	memory inputs to attention.
	The use of relative position is possible because sin(x+y) and cos(x+y) can be
	expressed in terms of y, sin(x) and cos(x).
	In particular, we use a geometric sequence of timescales starting with
	min_timescale and ending with max_timescale.  The number of different
	timescales is equal to channels / 2. For each timescale, we
	generate the two sinusoidal signals sin(timestep/timescale) and
	cos(timestep/timescale).  All of these sinusoids are concatenated in
	the channels dimension.
	Args:
	  length: scalar, length of timing signal sequence.
	  channels: scalar, size of timing embeddings to create. The number of
		  different timescales is equal to channels / 2.
	  min_timescale: a float
	  max_timescale: a float
	  start_index: index of first position
	Returns:
	  a Tensor of timing signals [1, length, channels]
	"""
	position = tf.cast(tf.range(length) + start_index, tf.float32)
	num_timescales = channels // 2
	log_timescale_increment = (
		tf.math.log(float(max_timescale) / float(min_timescale)) /
		(tf.cast(num_timescales, tf.float32) - 1))
	inv_timescales = min_timescale * tf.exp(
		tf.cast(tf.range(num_timescales), tf.float32) * -log_timescale_increment)
	scaled_time = tf.expand_dims(position, 1) * tf.expand_dims(inv_timescales, 0)
	signal = tf.concat([tf.sin(scaled_time), tf.cos(scaled_time)], axis=1)
	signal = tf.pad(signal, [[0, 0], [0, tf.math.mod(channels, 2)]])
	signal = tf.reshape(signal, [1, length, channels])
	return signal

def get_batch(batch_size, seq_len, n_feat):
	x_batch = np.zeros((batch_size, seq_len, n_feat))
	for i in range(batch_size):
		rand_pos = np.random.randint(0, high=seq_len) # random starting position.
		x_batch[i,:,:] = get_timing_signal_1d(seq_len, n_feat, start_index=rand_pos).numpy()
	y_batch = (-x_batch + 1.0)/2.0
	x_batch = (x_batch + 1.0)/2.0
	return x_batch, y_batch

if __name__ == '__main__':

	n_feat = 40
	batch_size = 3
	seq_len = 200

	x_train, y_train = get_batch(batch_size, seq_len, n_feat)
```

And here is what three training examples look like:

![image](https://user-images.githubusercontent.com/26111230/80768125-ddbc2a00-8b8c-11ea-9f93-60949c0b1562.png)

Binary cross-entropy or a squared error function could be used as the loss.

**Comment by jostosh** [comment]
Thanks for the suggestion. I'm afraid your example does not really resemble the typical maximum likelihood training scenario for which SPNs lend themselves well. Probably, one could still do something to similar to what you propose with SPNs, but it is to me not yet trivial how I should proceed. 

I'm working on getting dynamic SPNs to the forefront as we speak, but in the process, I stumbled on some false assumptions I made about the things I built so far and the compatibility with TensorFlow 2.x so it's taking a little bit longer than expected, but it should be there later ths week. I will include one of the datasets that's also used in the dynamic SPN paper.

**Comment by anicolson** [comment]
Sorry, I don't know what I was thinking lol. I can't believe I blanked on that. 

The "ozLevel" or "JapaneseVowlels" dataset from the Dynamic SPN paper looks good as the training time per iteration is low (unsure how big each dataset is though). 

Here is the Japanese vowels dataset: https://archive.ics.uci.edu/ml/datasets/Japanese+Vowels
Couldn't find the other one.

Here is the PenDigit dataset (i think): https://archive.ics.uci.edu/ml/datasets/Pen-Based+Recognition+of+Handwritten+Digits

Arabic digits:  https://archive.ics.uci.edu/ml/datasets/Spoken+Arabic+Digit
